### PR TITLE
define stake related actions and add handling logic

### PR DESCRIPTION
**Pull request by koseoyoung**
define 7 different action struct -- stakeCreate, stakeAdd, stakeAgain, stakeWithdraw, stakeUnstake, stakeChangeCandidate, stakeTransferOwnership

### CODE REVIEW

File: action/envelope.go [hunk 1]
```diff
@@ -88,6 +88,20 @@ func (elp *Envelope) Proto() *iotextypes.ActionCore {
...
+	case *StakeWithdraw:
+		actCore.Action = &iotextypes.ActionCore_StakeWithdraw{StakeWithdraw: act.Proto()}
+	case *StakeAdd:
+		actCore.Action = &iotextypes.ActionCore_StakeAddDeposit{StakeAddDeposit: act.Proto()}
+	case *StakeAgain:
+		actCore.Action = &iotextypes.ActionCore_StakeRestake{StakeRestake: act.Proto()}
+	case *StakeChangeCandidate:
+		actCore.Action = &iotextypes.ActionCore_StakeChangeCandidate{StakeChangeCandidate: act.Proto()}
+	case *StakeTransferOwnership:
+		actCore.Action = &iotextypes.ActionCore_StakeTransferOwnership{StakeTransferOwnership: act.Proto()}
```
Comment (CoderZhi) [inline action/envelope.go]: About naming:
Stake, Unstake, Restake, WithdrawStake, SwitchCandidate, TransferStake, DepositToStake

File: action/stakeagain.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+// IntrinsicGas returns the intrinsic gas of a StakeAgain
+func (sa *StakeAgain) IntrinsicGas() (uint64, error) {
+	payloadSize := uint64(len(sa.Payload()))
+	if (math.MaxUint64-StakeAgainBaseIntrinsicGas)/StakeAgainPayloadGas < payloadSize {
+		return 0, ErrOutOfGas
+	}
+
+	return payloadSize*StakeAgainPayloadGas + StakeAgainBaseIntrinsicGas, nil
```
Comment (CoderZhi) [inline action/stakeagain.go]: looks like a fix pattern, unify to a function
`func calculateIntrinsicGas(baseGas, gpu, size) (uint64, error)`

File: action/envelope.go (hunk 1 above)
Comment (koseoyoung) [reply action/envelope.go]: yes! renamed it

File: action/stakeagain.go (hunk 2 above)
Comment (koseoyoung) [reply action/stakeagain.go]: refactored in latest commit

File: action/deposittostake.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+// LoadProto converts a protobuf's Action to DepositToStake
+func (ds *DepositToStake) LoadProto(pbAct *iotextypes.StakeAddDeposit) error {
+	if pbAct == nil {
+		return errors.New("empty action proto to load")
+	}
+	ds = &DepositToStake{}
+
+	ds.bucketIndex = pbAct.GetBucketIndex()
+	ds.payload = pbAct.GetPayload()
```
Comment (dustinxie) [inline action/deposittostake.go]: if len(pbAct.GetAmount()) > 0?

Comment (dustinxie) [reply action/deposittostake.go]: corresponds to ds.amount != nil above

File: action/movestake.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+	}
+	sm = &MoveStake{}
+
+	sm.bucketIndex = pbAct.GetBucketIndex()
+	sm.payload = pbAct.GetPayload()
+	return nil
+}
+
+// SwitchCandidate defines the action of changing stake candidate ts the other
+type SwitchCandidate struct {
```
Comment (dustinxie) [inline action/movestake.go]: rename to ChangeCandidate, to be consistent with proto name

File: action/protocol/staking/protocol.go [hunk 5]
```diff
@@ -39,7 +39,22 @@ func NewProtocol() *Protocol {
...
-	//TODO
+	switch act := act.(type) {
+	case *action.Stake:
+		return p.handleStake(ctx, act, sm)
+	case *action.Unstake:
+		return p.handleUnstake(ctx, act, sm)
+	case *action.WithdrawStake:
+		return p.handleWithdrawStake(ctx, act, sm)
+	case *action.SwitchCandidate:
+		return p.handleSwitchCandidate(ctx, act, sm)
```
Comment (dustinxie) [inline action/protocol/staking/protocol.go]: handleChangeCandidate

File: action/reclaimstake.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+const (
+	// ReclaimStakePayloadGas represents the stake reclaim payload gas per uint
+	ReclaimStakePayloadGas = uint64(100)
+	// ReclaimStakeBaseIntrinsicGas represents the base intrinsic gas for stake reclaim
+	ReclaimStakeBaseIntrinsicGas = uint64(10000)
+)
+
+// ReclaimStake defines the action of stake restake/withdraw
+type ReclaimStake struct {
```
Comment (dustinxie) [inline action/reclaimstake.go]: this can be reclaimStake, since it is internal?

File: action/movestake.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+const (
+	// MoveStakePayloadGas represents the stake move payload gas per uint
+	MoveStakePayloadGas = uint64(100)
+	// MoveStakeBaseIntrinsicGas represents the base intrinsic gas for stake move
+	MoveStakeBaseIntrinsicGas = uint64(10000)
+)
+
+// MoveStake defines the action of changing stake candidate and transfering stake ownership
+type MoveStake struct {
```
Comment (dustinxie) [inline action/movestake.go]: this can be moveStake, since it is internal?

File: action/stake.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+const (
+	// StakePayloadGas represents the Stake payload gas per uint
+	StakePayloadGas = uint64(100)
+	// StakeBaseIntrinsicGas represents the base intrinsic gas for Stake create
+	StakeBaseIntrinsicGas = uint64(10000)
+)
+
+// Stake defines the action of Stake creation
+type Stake struct {
```
Comment (dustinxie) [inline action/stake.go]: rename to CreateStake?

File: action/stake.go [hunk 9]
```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2020 IoTeX Foundation
```
Comment (dustinxie) [inline action/stake.go]: createstake.go?

File: action/stake.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+func (sk *Stake) LoadProto(pbAct *iotextypes.StakeCreate) error {
+	if pbAct == nil {
+		return errors.New("empty action proto to load")
+	}
+	sk = &Stake{}
+
+	sk.candName = pbAct.GetCandidateName()
+	sk.duration = pbAct.GetStakedDuration()
+	sk.autoStake = pbAct.GetAutoStake()
+	sk.payload = pbAct.GetPayload()
```
Comment (dustinxie) [inline action/stake.go]: if len(pbAct.GetStakedAmount()) > 0
corresponds to sk.amount != nil above

**Pull request merged**